### ISSUE TITLE
scoop services install <app>

### ISSUE THREAD

**Issue by r15ch13** [lukesampson/scoop#1526]
Discussion about implementing service management similar to [`brew services`](https://github.com/Homebrew/homebrew-services).

Previous discussions about this topic: #1512, #513

I stumbled over [Laravel Valet for Windows](https://github.com/cretueusebiu/valet-windows) which uses [WinSW](https://github.com/cretueusebiu/valet-windows/blob/master/cli/Valet/WinSW.php) to setup a PHP developing environment. I know this is not the purpose of `scoop`, but I would like to create a documentation for setting up an updateable PHP dev environment. These zero-configuration setups (like xampp or valet) are cool, but most of them don't allow changing components or updates at all. With `scoop` it would be easy to swap out nginx for caddy/apache, MySQL for MariaDB and so on. Loosely coupled!

We could utilize https://github.com/kohsuke/winsw for installing/starting/stopping services.
In each manifest, we could provide a basic configuration that is copied to `~/scoop/persist/<app>/service.xml` so it can be easily modified.

E.g. `scoop services install nginx` would copy `winsw.exe` to `~/scoop/services/nginx.exe` (winsw has to be renamed) and register it, using a configuration located in `~/scoop/persist/nginx/service.xml`.
All other commands ` run|start|stop|restart` can be implemented like `brew services` commands.

Example installation for a full PHP environment:
```
λ scoop install php
λ scoop install composer
λ scoop install nginx
λ scoop install mariadb
λ scoop services install php-fpm
λ scoop services start php-fpm
λ scoop services install nginx
λ scoop services start nginx
```

Programs which could be run as a service:
- [apache](https://github.com/lukesampson/scoop/blob/master/bucket/apache.json)
- [caddy](https://github.com/lukesampson/scoop/blob/master/bucket/caddy.json)
- [minio](https://github.com/lukesampson/scoop/blob/master/bucket/minio.json)
- [php](https://github.com/lukesampson/scoop/blob/master/bucket/php.json)
- [rethinkdb](https://github.com/lukesampson/scoop/blob/master/bucket/rethinkdb.json)
- [unbound](https://github.com/lukesampson/scoop/blob/master/bucket/unbound.json)
- [mongodb](https://github.com/lukesampson/scoop/blob/master/bucket/mongodb.json)

What to you guys think?

**Comment by lukesampson** [comment]
WinSW sounds like it would be useful for programs that don't know how to run as a Windows service. Is that necessary though?

If it only needs to work with programs that *do* know how to run a service, the `sc.exe` commands are pretty straightforward, and they don't need the extra steps of copying the WinSW exe and writing XML.

| Scoop command | sc.exe equivalent |
|---|---|
| `scoop service install unbound` | `sudo sc.exe create Unbound binPath= "'$env:userprofile\scoop\apps\unbound\current\unbound.exe' -c '$env:userprofile\scoop\apps\unbound\current\service.conf' -w service" start= auto DisplayName= "Unbound DNS Validator"` |
| `scoop service start unbound` | `sudo sc.exe start Unbound` |
| `scoop service stop unbound` | `sudo sc.exe stop Unbound` |
| `scoop service restart unbound` | `sudo sc.exe stop Unbound; sudo sc.exe start Unbound` |

**Comment by r15ch13** [comment]
None of the programs I listed above can be registered as proper Windows services. They have to be started with WinSW or NSSM.

**Comment by lukesampson** [comment]
Did I screw something up in the Unbound example above? It seemed to work for me, and their offical Windows installer installs it as a service.

I would have thought any program suitable for use in a production environment on Windows would have support for running as a service.

And for running in development, would it be nicer to start a background process (as @nueko mentioned in #1513) so you didn't have to run as admin?

**Comment by r15ch13** [comment]
Okay unbound has build Windows services support.
But registering PHP with `sc.exe` results in a service hanging in `starting` state.

**Comment by deevus** [comment]
What about after a reboot?

**Comment by r15ch13** [comment]
Nope, PHP and the others don't support the Windows Service API. It can't be stopped or restarted properly.

See: https://stackoverflow.com/questions/10061191/add-nginx-exe-as-windows-system-service-like-apache/12964345#12964345

**Comment by tresf** [comment]
Perhaps use https://pecl.php.net/package/win32service.

**Comment by r15ch13** [comment]
That could be a solution for PHP yes. Leaves us with the rest :smile:

**Comment by lukesampson** [comment]
Roger that. You've convinced me that WinSW is the way to go.

**Comment by nueko** [comment]
Won't `sc.exe` hurt non-admin user?

**Comment by r15ch13** [comment]
WinSW won't work either if you don't have admin privileges.

**Comment by nueko** [comment]
just my thought `service start <app>` command can be run with Start-Process with hidden window

**Comment by nueko** [comment]
Just quick experiment

```powershell
param($cmd, $app, $argls)
. "$psscriptroot\..\lib\core.ps1"

# get Requested process
$requestedProcess = Get-Process $app -ErrorAction SilentlyContinue
$app, $process = $app.split('/')

function stop_process () {
    if ($requestedProcess) {
        Write-Host "Trying to gracefully exit..." -F Green

        # try gracefully first
        $requestedProcess.CloseMainWindow() | Out-Null

        # kill after five seconds
        Start-Sleep 5
        
        if (!$requestedProcess.HasExited) {
            Write-Host "Forcing stop process..." -F Yellow
            $requestedProcess | Stop-Process -Force
            $requestedProcess.Close();
        }
        else {
            Write-Host "Process $app sucessfully stoped" -F Green
        }
    }
}

function start_process () {
    if($argls) {
        Start-Process -WorkingDirectory $(appdir($app)) -FilePath ($process, $app)[!$process] -A $argls -WindowStyle Hidden -PassThru
    } else {
        Start-Process -WorkingDirectory $(appdir($app)) -FilePath ($process, $app)[!$process] -WindowStyle Hidden -PassThru
    }
}
switch ($cmd) {
    "stop" { stop_process }
    "start" { start_process }
    default { "scoop process: cmd '$cmd' not supported"; my_usage; exit 1 }
}

```

**Comment by gpetrov** [comment]
@r15ch13 and @lukesampson - what happen to this great services implementation? It looks amazing and it brings scoop to the level of Homebrew on mac, where you can also manage services directly.

So please finalize it.

**Comment by jtagcat** [comment]
I'd add syncthing to the list (easy to implement, and popular enough).

What I'd like to see is `scoop service enable/disable, enabling `start` on system startup.

**Comment by yarcowong** [comment]
So, what is this feature for now? I think the feature of start/stop server is enough for now. (without restarting the service when windows boot again). At least, we use windows for development purchase, must be rare to use it as a real server with services under *nix~
 And for now, it is a pain to start nginx which is installed by scoop~ I need to start it by hand and close whole powershell window if I want to restart it after change the config (not quite sure, there will be other solution)

**Comment by gpetrov** [comment]
Any new developments in the services support @lukesampson ?

**Comment by r15ch13** [comment]
Unfortunately, since I opened this issue there has not been any progress or other attempts to implement it.

(Luke is not active on this project)
